lm_eval/models/mlx_llms.py

```diff
@@ -0,0 +1,286 @@
+          by greedy sampling from the LM (that is, if the target string is the most likely N-token string to be output
+          by the LM given the input. )
+        """
+        try:
+            import mlx.core as mx
+        except ModuleNotFoundError:
+            raise Exception(
+                "attempted to use 'mlx' LM type, but package `mlx` is not installed. Please install mlx "
+                "via `pip install 'lm-eval[mlx]'` or `pip install -e '.[mlx]'`"
+            )
+        if not requests:
+            return []
+        res = []
+        # Keep order for later
+        original_order = {
+            (context, continuation): idx
+            for idx, (context, continuation) in enumerate(
+                [req.args for req in requests]
+            )
+        }
+
+        # sort the requests by their length (changes order)
+        idx = sorted(
+            range(len(requests)),
+            key=lambda i: len(requests[i].args[0]) + len(requests[i].args[1]),
+        )
+        if len(requests) < self.batch_size:
+            raise ValueError(
+                f"Dataset must have at least batch_size={self.batch_size}"
+                f" examples but only has {len(requests)}."
+            )
+
+        # Make the batches:
+        batch_idx = [
+            idx[i : i + self.batch_size]
+            for i in range(0, len(idx) - self.batch_size + 1, self.batch_size)
+        ]
+        if len(idx) % self.batch_size != 0:
+            # If the total requests size is not a multiple of the batch size, there will be a final
+            # batch with the remainder
+            batch_idx.append(idx[self.batch_size * len(batch_idx) :])
+        eval_logger.info(
+            f"{len(requests):,} requests and {len(batch_idx):,} {self.batch_size:,}-or-less-item batches"
+        )
+
+        # randomize the batches
+        indices = np.random.permutation(len(batch_idx))
+
+        pbar = tqdm(
+            total=len(indices),
+            disable=(disable_tqdm or (self.rank != 0)),
+            desc=f"Running loglikelihood requests ({len(batch_idx):,} batches)",
+        )
+        for batch_num, i in enumerate(indices):
+            context_batch = []
+            continuation_batch = []
+            full_sequences = []
+            prompt_lengths = []
+            for j in batch_idx[i]:
+                context, continuation = requests[j].args
+                context_batch.append(context)
+                continuation_batch.append(continuation)
+                prompt_lengths.append(len(self.tokenizer.encode(context, add_special_tokens = self.add_bos_token)))
+                full_sequence = self.tokenizer.encode(context + continuation, add_special_tokens = self.add_bos_token)
+                if full_sequence[-1] != self.tokenizer.eos_token_id:
+                    full_sequence.append(self.tokenizer.eos_token_id)
```

Comment:
why is this necessary?

Reply:
I'm keeping track of the context / prompt lengths for the mlx masking when calculating information about when target tokens are the same as the greedily-generated token (267 - 283)